File: desktop/src/renderer/state/marketplace-api-client.ts
// Typed fetch wrapper for the DestinCode marketplace Cloudflare Worker backend.
// Lives in renderer/ because the same React bundle runs on both desktop and Android —
// fetch calls go directly to the Worker (CORS allowlist covers both platforms).
// No IPC needed for read endpoints; write endpoints gate on token which callers supply via getToken().

export const MARKETPLACE_API_HOST = "https://destincode-marketplace-api.destinj101.workers.dev";

export class MarketplaceApiError extends Error {
  constructor(public readonly status: number, message: string) {
    super(message);
  }
}

export interface AuthStartResponse {
  device_code: string;
  user_code: string;
  auth_url: string;
  expires_in: number;
}

export type AuthPollResponse =
  | { status: "pending" }
  | { status: "complete"; token: string };

export interface StatsResponse {
  generated_at: number;
  plugins: Record<string, { installs: number; review_count: number; rating: number }>;
  themes: Record<string, { likes: number }>;
}

export interface PostRatingInput {
  plugin_id: string;
  stars: 1 | 2 | 3 | 4 | 5;
  review_text?: string;
}

export interface MarketplaceApiClient {
  getStats(): Promise<StatsResponse>;
  authStart(): Promise<AuthStartResponse>;
  authPoll(deviceCode: string): Promise<AuthPollResponse>;
  postInstall(pluginId: string): Promise<void>;
  postRating(input: PostRatingInput): Promise<{ hidden: boolean }>;
  deleteRating(pluginId: string): Promise<void>;
  toggleThemeLike(themeId: string): Promise<{ liked: boolean }>;
  postReport(input: { rating_user_id: string; rating_plugin_id: string; reason?: string }): Promise<void>;
}

export function createMarketplaceApiClient(opts: {
  host: string;
  getToken: () => string | null;
}): MarketplaceApiClient {
  const { host, getToken } = opts;

  async function request<T>(path: string, init: RequestInit & { auth?: boolean } = {}): Promise<T> {
    const headers: Record<string, string> = { "Content-Type": "application/json", ...(init.headers as Record<string, string>) };
    if (init.auth) {
      const token = getToken();
      if (!token) throw new MarketplaceApiError(401, "not signed in");
      headers.Authorization = `Bearer ${token}`;
    }
    // Remove the custom 'auth' flag before passing to fetch (not a standard RequestInit field)
    const { auth: _auth, ...fetchInit } = init;
    const res = await fetch(`${host}${path}`, { ...fetchInit, headers });
    // 202 Accepted is used for poll-pending responses — treat as success with { status: "pending" }
    const body = res.status === 202 ? { status: "pending" as const } : await res.json().catch(() => ({}));
    if (!res.ok && res.status !== 202) {
      throw new MarketplaceApiError(res.status, (body as { message?: string })?.message ?? res.statusText);
    }
    return body as T;
  }

  return {
    getStats: () => request<StatsResponse>("/stats", { method: "GET" }),
    authStart: () => request<AuthStartResponse>("/auth/github/start", { method: "POST" }),
    authPoll: (device_code) =>
      request<AuthPollResponse>("/auth/github/poll", {
        method: "POST",
        body: JSON.stringify({ device_code }),
      }),
    postInstall: async (plugin_id) => {
      await request("/installs", { method: "POST", body: JSON.stringify({ plugin_id }), auth: true });
    },
    postRating: (input) =>
      request<{ hidden: boolean }>("/ratings", {
        method: "POST",
        body: JSON.stringify(input),
        auth: true,
      }),
    deleteRating: async (plugin_id) => {
      await request(`/ratings/${encodeURIComponent(plugin_id)}`, { method: "DELETE", auth: true });
    },
    toggleThemeLike: (theme_id) =>
      request<{ liked: boolean }>(`/themes/${encodeURIComponent(theme_id)}/like`, {
        method: "POST",
        auth: true,
      }),
    postReport: async (input) => {
      await request("/reports", { method: "POST", body: JSON.stringify(input), auth: true });
    },
  };
}

File: app/src/main/kotlin/com/youcoded/app/marketplace/MarketplaceApiClient.kt
package com.youcoded.app.marketplace

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.net.URLEncoder
import java.util.concurrent.TimeUnit

/**
 * Kotlin mirror of the TS MarketplaceApiClient in desktop/src/renderer/state/marketplace-api-client.ts.
 * Calls the YouCoded Cloudflare Worker backend.
 *
 * WHY OkHttp: already a project dependency (for the WebSocket bridge server), so no new deps needed.
 * WHY suspend: all callers live in SessionService's serviceScope (Dispatchers.IO), consistent with
 * how other blocking I/O in that service is dispatched.
 */

private const val TAG = "MarketplaceApiClient"
private val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()

/** Mirrors the TS ApiResult<T> discriminated union — returned as JSON to the React renderer. */
sealed class ApiResult<out T> {
    data class Ok<T>(val value: T) : ApiResult<T>()
    data class Err(val status: Int, val message: String) : ApiResult<Nothing>()

    /** Serialize to a JSONObject matching the TS shape: { ok, value } or { ok, status, message } */
    fun toJson(serializeValue: (T) -> Any? = { v -> v }): JSONObject = when (this) {
        is Ok -> JSONObject().apply {
            put("ok", true)
            val v = serializeValue(value)
            when (v) {
                null               -> put("value", JSONObject.NULL)
                is JSONObject      -> put("value", v)
                is Boolean         -> put("value", v)
                is Int             -> put("value", v)
                is String          -> put("value", v)
                else               -> put("value", v.toString())
            }
        }
        is Err -> JSONObject().apply {
            put("ok", false)
            put("status", status)
            put("message", message)
        }
    }
}

class MarketplaceApiClient(
    private val store: MarketplaceAuthStore,
    private val host: String = "https://wecoded-marketplace-api.destinj101.workers.dev",
) {
    private val http = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    // ── Internal helpers ─────────────────────────────────────────────────────

    private suspend fun request(
        path: String,
        method: String = "GET",
        body: JSONObject? = null,
        auth: Boolean = false,
    ): Pair<Int, JSONObject> = withContext(Dispatchers.IO) {
        val reqBody = body?.toString()?.toRequestBody(JSON_MEDIA_TYPE)
        val builder = Request.Builder()
            .url("$host$path")
            .method(method, reqBody ?: if (method == "GET") null else "{}".toRequestBody(JSON_MEDIA_TYPE))
            .addHeader("Content-Type", "application/json")

        if (auth) {
            val token = store.getToken()
            // WHY: we don't log the token value — only whether it's present
            if (token == null) {
                return@withContext Pair(401, JSONObject().put("message", "not signed in"))
            }
            builder.addHeader("Authorization", "Bearer $token")
        }

        try {
            val resp = http.newCall(builder.build()).execute()
            val code = resp.code
            val raw = resp.body?.string() ?: "{}"
            // 202 Accepted = poll-pending — return synthetic pending body
            val json = if (code == 202) {
                JSONObject().put("status", "pending")
            } else {
                try { JSONObject(raw) } catch (_: Exception) { JSONObject() }
            }
            Pair(code, json)
        } catch (e: Exception) {
            Log.w(TAG, "HTTP $method $path failed: ${e.message}")
            Pair(0, JSONObject().put("message", e.message ?: "network error"))
        }
    }

    private fun errFromResponse(code: Int, body: JSONObject): ApiResult.Err =
        ApiResult.Err(code, body.optString("message", "HTTP $code"))

    // ── Public API (mirrors TS client method by method) ──────────────────────

    /** GET /stats — no auth required */
    suspend fun getStats(): ApiResult<JSONObject> {
        val (code, body) = request("/stats")
        return if (code == 200) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    /** POST /auth/github/start — initiates device-code OAuth flow */
    suspend fun authStart(): ApiResult<JSONObject> {
        val (code, body) = request("/auth/github/start", method = "POST")
        return if (code == 200) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    /**
     * POST /auth/github/poll — polls for token.
     * Returns either { status: "pending" } (202) or { status: "complete", token: "..." } (200).
     * WHY: on complete, the caller in SessionService saves the token to the store.
     */
    suspend fun authPoll(deviceCode: String): ApiResult<JSONObject> {
        val (code, body) = request(
            "/auth/github/poll",
            method = "POST",
            body = JSONObject().put("device_code", deviceCode),
        )
        return if (code == 200 || code == 202) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    /** POST /installs — records an install. Requires token. */
    suspend fun postInstall(pluginId: String): ApiResult<Unit> {
        val (code, body) = request(
            "/installs",
            method = "POST",
            body = JSONObject().put("plugin_id", pluginId),
            auth = true,
        )
        return if (code in 200..299) ApiResult.Ok(Unit) else errFromResponse(code, body)
    }

    /** POST /ratings — submit or update a rating. Requires token. */
    suspend fun postRating(
        pluginId: String,
        stars: Int,
        reviewText: String?,
    ): ApiResult<JSONObject> {
        val payload = JSONObject().apply {
            put("plugin_id", pluginId)
            put("stars", stars)
            if (!reviewText.isNullOrEmpty()) put("review_text", reviewText)
        }
        val (code, body) = request("/ratings", method = "POST", body = payload, auth = true)
        return if (code in 200..299) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    /** DELETE /ratings/:pluginId — remove the caller's rating. Requires token. */
    suspend fun deleteRating(pluginId: String): ApiResult<Unit> {
        val encoded = URLEncoder.encode(pluginId, "UTF-8")
        val (code, body) = request("/ratings/$encoded", method = "DELETE", auth = true)
        return if (code in 200..299) ApiResult.Ok(Unit) else errFromResponse(code, body)
    }

    /** POST /themes/:themeId/like — toggle like. Requires token. */
    suspend fun toggleThemeLike(themeId: String): ApiResult<JSONObject> {
        val encoded = URLEncoder.encode(themeId, "UTF-8")
        val (code, body) = request("/themes/$encoded/like", method = "POST", auth = true)
        return if (code in 200..299) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    /** POST /reports — report a rating. Requires token. */
    suspend fun postReport(
        ratingUserId: String,
        ratingPluginId: String,
        reason: String?,
    ): ApiResult<Unit> {
        val payload = JSONObject().apply {
            put("rating_user_id", ratingUserId)
            put("rating_plugin_id", ratingPluginId)
            if (!reason.isNullOrEmpty()) put("reason", reason)
        }
        val (code, body) = request("/reports", method = "POST", body = payload, auth = true)
        return if (code in 200..299) ApiResult.Ok(Unit) else errFromResponse(code, body)
    }
}
